### ISSUE TITLE
Task corrupts binary files

### ISSUE THREAD

**Issue by bodograumann** [qetza/vsts-replacetokens-task#193]
We naively used `targetFiles: '**/*'` which also included some ttf files.
Even though the binary files did not contain the tokenPrefix or tokenSuffix, `tfs-replace-tokens` still read them, decoded them and wrote them back encoded in some text-encoding, effectively corrupting them.

So I think the replacement process should:
- only write a file if replacements where made,
- encode and decode in such a way that the result without replacement is identical to the original and
- abort with an error if decoding was not possible.

We used the following parameters:
```
task: replacetokens@3
inputs:
  rootDirectory: '$(TargetFolder)'
  targetFiles: '**/*'
  encoding: 'auto'
  writeBOM: true
  actionOnMissing: 'warn'
  keepToken: false
  tokenPrefix: '#{'
  tokenSuffix: '}#'
  useLegacyPattern: false
  enableTelemetry: false
```

Here is an example, of how the ttf file was changed (after piping through `xxd`):
```
00000010: 8338 3ff2 0000 96f0 0000 004e 636d 6170  .8?....... | 00000010: ff38 3fff 0000 ffff 0000 004e 636d 6170  .8?.......
00000020: 8672 6003 0000 7df8 0000 03e0 6376 7420  .r`...}... | 00000020: ff72 6003 0000 7dff 0000 03ff 6376 7420  .r`...}...
00000030: d70b a375 0000 0110 0000 0198 6670 676d  ...u...... | 00000030: ff0b ff75 0000 0110 0000 01ff 6670 676d  ...u......
00000040: 8333 c24f 0000 00fc 0000 0014 676c 7966  .3.O...... | 00000040: ff33 ff4f 0000 00ff 0000 0014 676c 7966  .3.O......
00000050: 5275 ef99 0000 02d4 0000 71ea 6864 6d78  Ru........ | 00000050: 5275 ffff 0000 02ff 0000 71ff 6864 6d78  Ru........
```

**Comment by qetza** [comment]
Hi @bodograumann,
Thanks for reporting this issue.
The task is meant to be use on text file only so i never tested it on anything else. I will see if i can add some fail safe as you suggested: not writing file if no replace.
Decoding is always possible so the task cannot fail on decoding error. As for decode/encode which do not produce the same result if no replacement was done this may be due to the library used for decoding/encoding which is not meant for binary files.

**Comment by sergiusz-roszczyk** [comment]
Hi. I was hit on similar issue on one of my projects. It worked before and suddenly stopped working.
The issue is the same. Target files was set to `'**/*'` and on one project Java classes were corrupted with BOM char (despite file shouldn't be changed and on report 0 of 0 tokens were replaced).
On the other project one of index files in `.git` directory was corrupted. Again, 0 of 0 tokens replaced according to report.

So I fully agree that the intent is to change only text files but also only when any token was found inside the file and reported as changed.

**Comment by qetza** [comment]
I published a patch (3.7.1) with the following changes:
- if tokens where found: file is written (same as previous)
- if no tokens found but the output file is not the same as input file: file is copied (using binary read/write)
- else nothing is done

I've tested this with a binary file (JPG) and checksum before and after copy are the same.

**Comment by sergiusz-roszczyk** [comment]
Thanks @qetza . Azure DevOps already fetched new version. Initial tests look promising. I'll post some update over the weekend.
